controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/dataplanetoken/DataplaneTokenService.java

```diff
@@ -0,0 +1,121 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.restapi.dataplanetoken;
+
+import com.yahoo.config.provision.TenantName;
+import com.yahoo.security.token.Token;
+import com.yahoo.security.token.TokenCheckHash;
+import com.yahoo.security.token.TokenDomain;
+import com.yahoo.security.token.TokenGenerator;
+import com.yahoo.transaction.Mutex;
+import com.yahoo.vespa.hosted.controller.Controller;
+import com.yahoo.vespa.hosted.controller.api.integration.dataplanetoken.DataplaneToken;
+import com.yahoo.vespa.hosted.controller.api.integration.dataplanetoken.DataplaneTokenVersions;
+import com.yahoo.vespa.hosted.controller.api.integration.dataplanetoken.FingerPrint;
+import com.yahoo.vespa.hosted.controller.api.integration.dataplanetoken.TokenId;
+import com.yahoo.vespa.hosted.controller.persistence.CuratorDb;
+
+import java.nio.charset.StandardCharsets;
+import java.security.Principal;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Service to list, generate and delete data plane tokens
+ *
+ * @author mortent
+ */
+public class DataplaneTokenService {
+
+    private static final String TOKEN_PREFIX = "vespa_cloud_";
+    private static final byte[] FINGERPRINT_CONTEXT = new byte[0];
+    private static final int TOKEN_BYTES = 16;
+
+
+    private final Controller controller;
+
+    public DataplaneTokenService(Controller controller) {
+        this.controller = controller;
+    }
+
+    /**
+     * List valid tokens for a tenant
+     */
+    public List<DataplaneTokenVersions> listTokens(TenantName tenantName) {
+        return controller.curator().readDataplaneTokens(tenantName);
+    }
+
+    /**
+     * Generates a token using tenant name as the check access context.
+     * Persists the token fingerprint and check access hash, but not the token value
+     *
+     * @param tenantName name of the tenant to connect the token to
+     * @param tokenId The user generated name/id of the token
+     * @param principal The principal making the request
+     * @return a DataplaneToken containing the secret generated token
+     */
+    public DataplaneToken generateToken(TenantName tenantName, TokenId tokenId, Principal principal) {
+        TokenDomain tokenDomain = new TokenDomain(FINGERPRINT_CONTEXT, tenantName.value().getBytes(StandardCharsets.UTF_8));
+        Token token = TokenGenerator.generateToken(tokenDomain, TOKEN_PREFIX, TOKEN_BYTES);
+        TokenCheckHash checkHash = TokenCheckHash.of(token, TOKEN_BYTES);
```

Comment:
We should probably have a separate `CHECK_HASH_BYTES` for this. I think it would make sense for check hashes to be 32 bytes, as they are not really intended for human consumption.

Reply:
Also, should `TOKEN_BYTES` be 32 as well? This would be 256 bits of CSPRNG randomness per token.